pkg/network/bootstrap/bootstrap.go

```diff
@@ -0,0 +1,331 @@
+		}
+		defer func() {
+			if err := probe.Stop(); err != nil {
+				log.Error().Err(err).Msg("could not stop DHCP probe properly")
+			}
+		}()
+
+		var addrs4 = newAddrSet()
+		var addrs6 = newAddrSet()
+		for stay := true; stay; {
+			select {
+			case <-ctx.Done():
+				// exit for loop
+				stay = false
+				break
+
+			default:
+				addrs, err := netlink.AddrList(link, netlink.FAMILY_V4)
+				if err != nil {
+					return errors.Wrapf(err, "could not read address from interface %s", name)
+				}
+				addrs4.AddSlice(addrs)
+
+				addrs, err = netlink.AddrList(link, netlink.FAMILY_V6)
+				if err != nil {
+					return errors.Wrapf(err, "could not read address from interface %s", name)
+				}
+				addrs6.AddSlice(addrs)
+
+				if addrs6.Len() > 0 && addrs4.Len() > 0 {
+					stay = false
+					break
+				}
+				time.Sleep(time.Second)
+			}
+		}
+
+		routes4, err := netlink.RouteList(link, netlink.FAMILY_V4)
+		if err != nil {
+			return errors.Wrapf(err, "could not read routes from interface %s", name)
+		}
+		routes6, err := netlink.RouteList(link, netlink.FAMILY_V6)
+		if err != nil {
+			return errors.Wrapf(err, "could not read routes from interface %s", name)
+		}
+
+		cAddrs <- IfaceConfig{
+			Name:    name,
+			Addrs4:  addrs4.ToSlice(),
+			Addrs6:  addrs6.ToSlice(),
+			Routes4: routes4,
+			Routes6: routes6,
+		}
+
+		return nil
+	})
+	if err != nil {
+		log.Info().Str("interface", link.Attrs().Name).Msg("failed to gather ip addresses")
+	}
+
+	return nil
+}
+
+// SelectZOS decide which interface should be assigned to the ZOS bridge
+// if multiple interfaces receives an IP from DHCP
+// we prefer a interfaces that has the smallest IP and private IP gateway
+// if none is found, then we pick the interface that has the smallest IP and any IP gateway
+func SelectZOS(cfgs []IfaceConfig) (string, error) {
+
+	selected4 := cfgs[:0]
+	// selected6 := cfgs[:0]
+	for _, cfg := range cfgs {
+		for _, route := range cfg.Routes4 {
+			if route.Gw != nil && isPrivateIP(route.Gw) {
+				selected4 = append(selected4, cfg)
+			}
+		}
+
+		// for _, route := range cfg.Routes4 {
+		// 	if route.Gw != nil && isPrivateIP(route.Gw) {
+		// 		selected6 = append(selected6, cfg)
+		// 	}
+		// }
+	}
+
+	if len(selected4) < 1 {
+		for _, cfg := range cfgs {
+			for _, route := range cfg.Routes4 {
+				if route.Gw != nil {
+					selected4 = append(selected4, cfg)
+				}
+			}
+
+			// for _, route := range cfg.Routes6 {
+			// 	if route.Gw != nil {
+			// 		selected6 = append(selected6, cfg)
+			// 	}
+			// }
+		}
+	}
+
+	if len(selected4) < 1 {
+		return "", fmt.Errorf("not route with default gateway found")
+	}
+
+	sort.Sort(byIP4(selected4))
+
+	return selected4[0].Name, nil
+}
+
+type ifaceFilter func(links []netlink.Link) []netlink.Link
+
+// filterPhysical is a ifaceFilter that filter out the links that are
+// not physical interface
+func filterPhysical(links []netlink.Link) []netlink.Link {
+	out := links[:0]
```

Comment:
Why this slicing if it will create an empty slice anyway? 

```suggestion
	var out []netlink.Link
```

Reply:
This is not the same, the idea here is to have a slice that points to the same underlying array so we can filter without copying. See https://github.com/golang/go/wiki/SliceTricks#filtering-without-allocating